File: python/algorithms/edmonds_karp.py
import numpy as np
from typing import List


def edmonds_karp(capacity: np.ndarray, s: int, t: int) -> float:
    # Edmonds Karp maximum flow algorithm
    V = len(capacity)
    flow = 0
    residual = np.zeros((V, V))

    while True:
        # BFS
        q = [s]
        p = np.zeros(V, dtype=int) - 1
        df = float("inf")
        while q:
            u = q.pop(0)
            for v in range(V):
                if capacity[u, v] > 0 and capacity[u, v] > residual[u, v] and p[v] == -1:
                    p[v] = u
                    df = min(df, capacity[u, v] - residual[u, v])
                    if v != t:
                        q.append(v)

        if p[t] >= 0:
            flow += df
            v = t
            while v != s:
                u = p[v]
                residual[u, v] += df
                residual[v, u] -= df
                v = u
        else:
            break

    return flow

File: src/algorithms/edmonds_karp.rs
use ndarray::{Array1, Array2, ArrayView2, Axis};
use std::collections::VecDeque;


pub fn edmonds_karp(capacity: &ArrayView2<'_, f64>, s: usize, t: usize) -> f64 {
    let nv: usize = capacity.len_of(Axis(0));
    let mut flow: f64 = 0.;
    let mut residual = Array2::zeros((nv, nv));
    
    loop {
        // BFS
        // BFS queue
        let mut q: VecDeque<usize> = VecDeque::new();
        q.push_back(s);
        // parent vertex in shortest path
        // initialize to `nv` to represent null
        let mut p = Array1::zeros::<usize>(nv);
        p.iter_mut().for_each(|x| *x = nv);
        let mut df = f64::MAX;
        while q.len() > 0 {
            let u = q.pop_front().unwrap();
            for v in 0..nv {
                let cap: f64 = capacity[[u, v]];
                let res: f64 = residual[[u, v]];
                if cap > 0. && cap > res && p[v] == nv {
                    p[v] = u;
                    let _df = cap - res;
                    if _df < df {
                        df = _df;
                    }
                    if v != t {
                        q.push_back(v);
                    }
                }
            }
        }

        if p[t] < nv {
            flow += df;
            let mut v = t;
            while v != s {
                let u = p[v];
                residual[[u, v]] += df;
                residual[[v, u]] -= df;
                v = u;
            }
        }
        else {
            break
        }
    }
    
    flow
}